src/selectors/getCurrentProjectInstructions.js

```diff
@@ -1,9 +1,24 @@
 import {createSelector} from 'reselect';
 import getCurrentProjectKey from './getCurrentProjectKey';
+import getCurrentProjectInstructionsUnsaved
+  from './getCurrentProjectInstructionsUnsaved';
+import isEditingInstructions from './isEditingInstructions';
 import getProjects from './getProjects';
 
 export default createSelector(
-  [getCurrentProjectKey, getProjects],
-  (projectKey, projects) =>
-    projectKey ? projects.getIn([projectKey, 'instructions']) : '',
+  [
+    getCurrentProjectKey, getProjects,
+    isEditingInstructions, getCurrentProjectInstructionsUnsaved,
+  ],
+  (projectKey, projects, isEditing, instructionUnsaved) => {
+    if (!projectKey) {
+      return '';
+    }
+
+    if (isEditing && instructionUnsaved) {
+      return instructionUnsaved;
```

Comment:
I did add getCurrentProjectinstructionsUnsaved, and note that this selector invokes that one, but I did like that this was all pulled when the HOC is composed like this:
```js
function mapStateToProps(state) {
  return {
    instructions: getCurrentProjectInstructions(state),
    isEditing: isEditingInstructions(state),
    isOpen: !getHiddenUIComponents(state).includes('instructions'),
    projectKey: getCurrentProjectKey(state),
  };
}

```

So this selector sort of becomes the router on which instructions to pull, without having to change the existing logic or add another wrapper on top of those two. I think if this needs to get any more complex in the future it would make sense to break it out, but for now I added some comments explaining. If you still want me to update it, happy to do it 👍

Reply:
I feel you on the appeal of having a “router” selector for sure. My concern is just making _this_ selector that router—the name implies the more straightforward previous behavior. An example of where this would be an issue is if `getCurrentProjectInstructions` were used somewhere else in the application to decide whether to display a “Delete Instructions” button (based on the selector returning a truthy value). This would yield weird behavior as currently written.

So, I would propose creating a new selector that composes `getCurrentProjectInstructions` and `getCurrentProjectInstructionsUnsaved` (although again I would go for something like `getCurrentProjectDraftInstructions`), and using that here when we explicitly want the “routing behavior”.

That said, I am also not _totally_ convinced we want to be routing the unsaved instructions through the `instructions` prop in the `<Instructions>` component. Might it be simpler to reason about if the component were explicitly given the instructions for display and the draft instructions for the editor as separate props? In fact, it seems like a `draftInstructions` prop could also do the job of `isEditing`, i.e. the mere presence of `draftInstructions` would be the signal that we are currently editing.

WDYT?

Reply:
All good points. I would be fine with putting it in the render of the `<Instructions>` component and passing both values in. It's really just a one-liner.